execution-engine/src/native_modules/tflite_inference.rs

```diff
@@ -0,0 +1,183 @@
+//! A native module for ML inference on TensorFlow Lite.
+//! Takes an input tensor, feeds it to the model and outputs an output tensor.
+//!
+//! ## Authors
+//!
+//! The Veracruz Development Team.
+//!
+//! ## Licensing and copyright notice
+//!
+//! See the `LICENSE_MIT.markdown` file in the Veracruz root directory for
+//! information on licensing and copyright.
+
+use crate::fs::{FileSystem, FileSystemResult};
+use crate::native_modules::common::Service;
+use lazy_static::*;
+use libc::c_int;
+use serde::Deserialize;
+use std::env::{current_dir, set_current_dir};
+use std::fs::{create_dir, create_dir_all, remove_dir_all, File};
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+use tflite::ops::builtin::BuiltinOpResolver;
+use tflite::{FlatBufferModel, InterpreterBuilder};
+use wasi_types::ErrNo;
+
+lazy_static! {
+    // Current directory before moving to the module's directory
+    static ref WORK_DIR: PathBuf = current_dir().unwrap_or(PathBuf::from(""));
+
+    // The native module executes in a dedicated directory to mitigate data
+    // leaks
+    static ref MODULE_DIR: PathBuf = WORK_DIR.clone().join("tflite_inference_service");
+}
+
+/// Module's API.
+#[derive(Deserialize, Debug)]
+pub(crate) struct TfLiteInferenceService {
+    // TODO: support several inputs and outputs
+    /// Path to the input tensor to be fed to the network.
+    input_tensor_path: PathBuf,
+    /// Path to the model serialized with FlatBuffers.
+    model_path: PathBuf,
+    /// Path to the output tensor containing the result of the prediction.
+    output_tensor_path: PathBuf,
+    /// Number of CPU threads to use for the TensorFlow Lite interpreter.
+    num_threads: c_int,
+}
+
+impl Service for TfLiteInferenceService {
+    /// Return the name of this service
+    fn name(&self) -> &str {
+        "TensorFlow Lite Inference Service"
+    }
+
+    /// Triggers the service. The details of the service can be found in
+    /// function `infer`.
+    /// Here is the enter point. It also erases the state unconditionally
+    /// afterwards.
+    fn serve(&mut self, fs: &mut FileSystem, _input: &[u8]) -> FileSystemResult<()> {
+        // when reaching here, the `input` bytes are already parsed.
+
+        // Prepare module's directory
+        create_dir(MODULE_DIR.as_path()).map_err(|_| ErrNo::Access)?;
+        set_current_dir(MODULE_DIR.as_path())?;
+
+        let result = self.infer(fs);
+        // NOTE: erase all the states and files passed to the module and created
+        // during execution
+        self.reset();
+        result
+    }
+
+    /// For the purpose of demonstration, we always return true. In reality,
+    /// this function may check validity of the `input`, and even buffer the
+    /// result for further uses.
```

Comment:
Since this is no longer just a demo, should we change this?

Reply:
Sure. Should we perform any check on the input? We can't really access any sensitive file since the module attempts to read them from the VFS with the program's permissions, before copying them to the FS.

Reply:
I can't see any trivial way to inject malformed input here. Let's just amend the comment

Reply:
Should the number of threads be configurable by the caller? Can you think of any attacks leveraging that?